homework10/task1/matrix.c

```diff
@@ -0,0 +1,183 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include "list.h"
+#include "matrix.h"
+
+typedef struct CityData {
+    int cities;
+    int roads;
+    int** matrix;
+    int capitalsAmount;
+    List** citiesOfCountry;
+} CityData;
+
+CityData* createCityData() {
+    return calloc(1, sizeof(CityData));
+}
+
+CityData* readDataFromFile(FILE* file) {
+    CityData* data = malloc(sizeof(CityData));
+    if (!data) {
+        return NULL;
+    }
+
+    fscanf(file, "%d", &data->cities);
+    fscanf(file, "%d", &data->roads);
+
+    data->matrix = calloc(data->cities + 1, sizeof(int*));
+    for (int i = 1; i <= data->cities; ++i) {
+        data->matrix[i] = calloc(data->cities + 1, sizeof(int));
+    }
+
+    for (int i = 0; i < data->roads; ++i) {
+        int row = 0;
+        int column = 0;
+        int length = 0;
+        fscanf(file, "%d %d %d", &row, &column, &length);
+        data->matrix[row][column] = data->matrix[column][row] = length;
+    }
+
+    fscanf(file, "%d", &data->capitalsAmount);
+    data->citiesOfCountry = calloc(data->cities + 1, sizeof(List*));
+    if (data->citiesOfCountry == NULL) {
+        free(data);
+        return NULL;
+    }
+
+    bool* marked = calloc(data->cities + 1, sizeof(bool));
+    for (int i = 0; i < data->capitalsAmount; i++) {
+        int capital = 0;
+        fscanf(file, "%d", &capital);
+        data->citiesOfCountry[capital] = createList();
+        insert(capital, data->citiesOfCountry[capital]);
+        marked[capital] = true;
+    }
+
+    free(marked);
+
+    return data;
+}
+
+void conquer(List* state, int** matrix, bool marked[], int cities) {
+    int nearestDistance = INT_MAX;
+    int nearestCity = -1;
+    int length = listLength(state);
+    for (int i = 0; i < length; ++i) {
+        int stateCity = accessElement(i, state);
+        for (int j = 1; j <= cities; ++j) {
+            int matrixElement = matrix[stateCity][j];
+            if (matrixElement != 0 && marked[j] != true) {
```

Comment:
```suggestion
            if (matrixElement != 0 && !marked[j]) {
```
Сравниваться с булевыми константами запрещено стайлгайдом, это какое-то неуважение к булевой алгебре :)